src/store/wallet/sagas.ts

```diff
@@ -29,22 +34,47 @@ const errors = {
 }
 
 function* sendAsync({
-  payload: { recipientWallet, amount }
+  payload: { recipientWallet, amount, chain }
 }: ReturnType<typeof send>) {
   const account = yield select(getAccountUser)
   const weiBNAmount = stringWeiToBN(amount)
-  const weiBNBalance: ReturnType<typeof getAccountBalance> = yield select(
-    getAccountBalance
+  const weiBNBalance: BNWei = yield select(getAccountBalance) ??
+    (new BN('0') as BNWei)
+
+  const waudioWeiAmount: BNWei = yield call(
+    walletClient.getCurrentWAudioBalance
   )
-  if (!weiBNBalance || !weiBNBalance.gte(weiBNAmount)) return
+  if (chain === Chain.Eth && (!weiBNBalance || !weiBNBalance.gt(weiBNAmount))) {
+    yield put(sendFailed({ error: 'Not enough $AUDIO' }))
+    return
+  } else if (chain === Chain.Sol) {
+    const totalBalance = waudioWeiAmount.add(weiBNBalance)
+    if (weiBNAmount.gt(totalBalance)) {
+      yield put(sendFailed({ error: 'Not enough $AUDIO' }))
+      return
+    }
+  }
+
   try {
     yield put(
       make(Name.SEND_AUDIO_REQUEST, {
         from: account?.wallet,
         recipient: recipientWallet
       })
     )
-    yield call(() => walletClient.sendTokens(recipientWallet, weiBNAmount))
+    if (chain === Chain.Sol && weiBNAmount.gt(waudioWeiAmount)) {
+      // transfer all eth AUDIO to WAUDIO
+      yield put(transferingEthAudioToSolWAudio())
+      yield call(walletClient.transferTokensFromEthToSol)
+    }
+
+    if (chain === Chain.Eth) {
+      yield call(() => walletClient.sendTokens(recipientWallet, weiBNAmount))
+    } else {
+      yield call(() =>
+        walletClient.sendWAudioTokens(recipientWallet, weiBNAmount)
+      )
```

Comment:
The branching logic in this function is quite confusing imo. Could we refactor this to have some top level branch on chain type, and then the various amount conditions within that?